src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -62,5 +63,68 @@ public abstract class PipeReader
         /// Cancel the pending <see cref="ReadAsync"/> operation. If there is none, cancels next <see cref="ReadAsync"/> operation, without completing the <see cref="PipeWriter"/>.
         /// </summary>
         public abstract void OnWriterCompleted(Action<Exception, object> callback, object state);
+
+        /// <summary>
+        /// Asynchronously reads the bytes from the <see cref="PipeReader"/> and writes them to the specified stream, using a specified buffer size and cancellation token.
+        /// </summary>
+        /// <param name="destination">The stream to which the contents of the current stream will be copied.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        public virtual async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                SequencePosition consumed = default;
+
+                try
+                {
+                    ReadResult result = await ReadAsync(cancellationToken).ConfigureAwait(false);
+                    ReadOnlySequence<byte> buffer = result.Buffer;
+
+                    consumed = buffer.Start;
+
+                    if (result.IsCanceled)
+                    {
+                        throw new OperationCanceledException();
+                    }
+
+                    if (!buffer.IsEmpty)
+                    {
+                        if (buffer.IsSingleSegment)
+                        {
+                            await destination.WriteAsync(buffer.First, cancellationToken).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            SequencePosition position = buffer.Start;
+                            while (buffer.TryGet(ref position, out ReadOnlyMemory<byte> memory))
```

Comment:
Is IsSingleSegment + First measurably faster than Start + TryGet?  I'm wondering if the first if blocks needed.

Reply:
I dunno if it'll matter for this case, I haven't measured.